test/controllers/works_controller_test.rb

```diff
@@ -22,95 +33,254 @@
 
+      expect {
+        patch work_path(id), params: work_hash
+      }.wont_change 'Work.count'
+
+      must_respond_with :redirect
+      must_redirect_to work_path(id)
+
+      new_work = Work.find_by(id: id)
+
+      expect(new_work.title).must_equal work_hash[:work][:title]
+      expect(new_work.creator).must_equal work_hash[:work][:creator]
+      expect(new_work.description).must_equal work_hash[:work][:description]
     end
 
     it "renders bad_request for bogus data" do
+      work_hash[:work][:title] = nil
+      id = works(:poodr).id
+      old_poodr = works(:poodr)
+
 
+      expect {
+        patch work_path(id), params: work_hash
+      }.wont_change 'Work.count'
+      new_poodr = Work.find(id)
+
+      must_respond_with :bad_request
+      expect(old_poodr.title).must_equal new_poodr.title
+      expect(old_poodr.creator).must_equal new_poodr.creator
+      expect(old_poodr.description).must_equal new_poodr.description
     end
 
     it "renders 404 not_found for a bogus work ID" do
+      id = -1
+
+      expect {
+        patch work_path(id), params: work_hash
+      }.wont_change 'Work.count'
+
+      must_respond_with :not_found
 
     end
   end
 
   describe "destroy" do
-    it "succeeds for an extant work ID" do
+    it "succeeds for an existing work ID" do
+      id = works(:poodr).id
+      title = works(:poodr).title
+
+      # Act - Assert
+      expect {
+        delete work_path(id)
+      }.must_change 'Work.count', -1
+
+      must_respond_with :redirect
+      must_redirect_to root_path
 
+      expect(Work.find_by(id: id)).must_equal nil
     end
 
     it "renders 404 not_found and does not update the DB for a bogus work ID" do
+      id = -1
+      expect {
+        delete work_path(id)
+        # }.must_change 'Book.count', 0
+      }.wont_change 'Work.count'
+
+      must_respond_with :not_found
 
     end
   end
 
   describe "upvote" do
-
     it "redirects to the work page if no user is logged in" do
+      id = works(:poodr).id
 
-    end
+      expect{
+        post upvote_path(id)}.wont_change 'Vote.count'
 
-    it "redirects to the work page after the user has logged out" do
+        must_redirect_to work_path(id)
+        # expect(flash[:warning]).must_equal "You must be logged in to vote for a work."
 
     end
 
-    it "succeeds for a logged-in user and a fresh user-vote pair" do
 
-    end
+      it "redirects to the work page after the user has logged out" do
+      end
 
-    it "redirects to the work page if the user has already voted for that work" do
+      it "succeeds for a logged-in user and a fresh user-vote pair" do
+        user = users(:grace)
+        id = works(:poodr).id
+        vote = Vote.new(user: @login_user, work: @work)
 
-    end
+        perform_login(user)
+
+        post upvote_path(id)
+
+        must_redirect_to work_path(id)
+
+      end
+
+      it "redirects to the work page if the user has already voted for that work" do
+        user = users(:grace)
```

Comment:
This is a pretty good attempt at trying to test the login functionality along with voting.

Problems:  1.  Your sessions controller isn't setting `sessions[:user_id]`
2.  Somehow the `post upvote_path(id)` is clearing session completely erasing everything in it.  I'm not sure how.